File: static_files/assignments/HW2_files/SRC/Python/RDM_maker.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import savemat

def RDM_maker(data, region, data_type):
    """
    Generate Representational Dissimilarity Matrix (RDM) from neural data.
    
    Parameters:
    - data: 3D numpy array [n_neurons x n_stimuli x n_times]
    - region: string name of the brain region
    - data_type: string representing data type
    
    Returns:
    - RDM: 3D numpy array [n_stimuli x n_stimuli x n_times]
    """

    n_neurons, n_stimuli, n_times = data.shape
    RDM = np.zeros((n_stimuli, n_stimuli, n_times))

    for t in range(n_times):
        slice_data = data[:, :, t]  # shape: [n_neurons x n_stimuli]

        for i in range(n_stimuli):
            for j in range(n_stimuli):
                cor = np.corrcoef(slice_data[:, i], slice_data[:, j])[0, 1]
                RDM[i, j, t] = 1 - cor  # dissimilarity

    # Visualize one time slice (e.g., time index 60)
    plt.imshow(RDM[:, :, 59], cmap='hot', interpolation='nearest')
    plt.title(f'RDM at time slice 60')
    plt.colorbar()
    plt.show()

    # Save the RDM to .mat file
    savemat(f'RDM_Matrix_{data_type}_{region}.mat', {'RDM': RDM})

    return RDM

File: static_files/assignments/HW2_files/SRC/MATLAB/RDM_maker.m
function RDM = RDM_maker(data , region , dataType)


% data: 3D array [nNeurons x nStimuli x nTimes] = (858 x 144 x 171)
[nNeurons, nStimuli, nTimes] = size(data);

% Initialize the RDM array: [144 x 144 x 171]
RDM = zeros(nStimuli, nStimuli, nTimes);

for t = 1:nTimes
    % Extract data across all neurons/stimuli for time slice t
    % sliceData is [858 x 144]
    sliceData = data(:,:,t);
    
    % Compute correlation among stimuli (columns)
    % corr(sliceData) returns a 144 x 144 correlation matrix
    
    for i = 1: size(sliceData,2)
        for j = 1: size(sliceData,2)
            corMat = corr(sliceData(:,i) , sliceData(:,j));
    
            % Convert correlation to dissimilarity = 1 - correlation
            RDM(i,j,t) = 1 - corMat;
        end
    end
end
% imagesc for a sample time slice
imagesc(RDM(i,j,60));

save(['RDM_Matrix_',dataType ,'_', region, '.mat'], 'RDM', '-v7.3');
end


